slither/detectors/operations/cache_array_length.py

```diff
@@ -0,0 +1,215 @@
+
+        # Array length may change in the following situations:
+        # - when `push` is called
+        # - when `pop` is called
+        # - when `delete` is called on the entire array
+        if node.type == NodeType.EXPRESSION:
+            for op in node.irs:
+                if isinstance(op, Length) and op.value == array:
+                    # op accesses array.length, not necessarily modifying it
+                    return True
+                if isinstance(op, Delete):
+                    # take into account only delete entire array, since delete array[i] doesn't change `array.length`
+                    if (
+                        isinstance(op.expression, UnaryOperation)
+                        and isinstance(op.expression.expression, Identifier)
+                        and op.expression.expression.value == array
+                    ):
+                        return True
+
+        for son in node.sons:
+            if son not in visited:
+                if CacheArrayLength._is_loop_referencing_array_length(son, visited, array, depth):
+                    return True
+        return False
+
+    @staticmethod
+    def _handle_loops(nodes: List[Node], non_optimal_array_len_usages: List[SourceMapping]) -> None:
+        """
+        For each loop, checks if it has a comparison with `length` array member and, if it has, checks whether that
+        array size could potentially change in that loop.
+        If it cannot, the loop condition is added to `non_optimal_array_len_usages`.
+        There may be some false negatives here - see docs for `_is_loop_referencing_array_length` for more information.
+        """
+        for node in nodes:
+            if node.type == NodeType.STARTLOOP:
+                if_node = node.sons[0]
+                if if_node.type != NodeType.IFLOOP:
+                    continue
+                if not isinstance(if_node.expression, BinaryOperation):
+                    continue
+                exp: BinaryOperation = if_node.expression
+                if not CacheArrayLength._is_identifier_member_access_comparison(exp):
+                    continue
+                array: StateVariable
+                if isinstance(exp.expression_right, MemberAccess):
+                    array = CacheArrayLength._extract_array_from_length_member_access(
+                        exp.expression_right
+                    )
+                else:  # isinstance(exp.expression_left, MemberAccess) == True
+                    array = CacheArrayLength._extract_array_from_length_member_access(
+                        exp.expression_left
+                    )
+                if array is None:
+                    continue
+
+                visited: Set[Node] = set()
+                if not CacheArrayLength._is_loop_referencing_array_length(
+                    if_node, visited, array, 1
+                ):
+                    non_optimal_array_len_usages.append(if_node.expression)
```

Comment:
Instead of tracking the expression, if we track the node (`Node`), it will render properly

Reply:
Will change it in the next PR. 

However, the small issue here is that Slither will output the `Node` using internal variable names. So, for example, it may output something like: `Loop condition at i_scope_4 < array.length [...]`, for the variable `i`, which may be a bit misleading for users.

Reply:
I think that's acceptable although not ideal. Other detectors do this as well